Source/Core/InputCommon/ImageOperations.cpp

```diff
@@ -0,0 +1,381 @@
+finalise:
+  if (info_ptr != nullptr)
+    png_free_data(png_ptr, info_ptr, PNG_FREE_ALL, -1);
+  if (png_ptr != nullptr)
+    png_destroy_write_struct(&png_ptr, (png_infopp) nullptr);
+
+  return success;
+}
+
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+ImagePixelData MaskColor(const ImagePixelData& src, const Pixel& color)
+{
+  ImagePixelData result(src.width, src.height, Pixel{255, 255, 255, 255});
+  for (u32 x = 0; x < src.width; x++)
+  {
+    for (u32 y = 0; y < src.height; y++)
+    {
+      const auto pixel = src.pixels[y * src.width + x];
+      if (pixel == color)
+      {
+        result.pixels[y * src.width + x] = Pixel{0, 0, 0, 255};
+      }
+    }
+  }
+  return result;
+}
+
+ImagePixelData Resize(ResizeMode mode, const ImagePixelData& src, u32 new_width, u32 new_height)
+{
+  ImagePixelData result;
+  result.width = new_width;
+  result.height = new_height;
+  result.pixels.resize(new_width * new_height);
+
+  for (u32 x = 0; x < new_width; x++)
+  {
+    const double u = x / static_cast<double>(new_width - 1);
+    for (u32 y = 0; y < new_height; y++)
+    {
+      const double v = y / static_cast<double>(new_height - 1);
+
+      switch (mode)
+      {
+      case Nearest:
+        result.pixels[y * new_width + x] = SampleNearest(src, u, v);
+        break;
+      }
+    }
+  }
+
+  return result;
+}
+
+ImagePixelData ResizeKeepAspectRatio(ResizeMode mode, const ImagePixelData& src, u32 new_width,
+                                     u32 new_height, const Pixel& background_color)
+{
+  ImagePixelData result(new_width, new_height, background_color);
+
+  const double corrected_height = new_width * (src.height / static_cast<double>(src.width));
+  const double corrected_width = new_height * (src.width / static_cast<double>(src.height));
+  // initially no borders
+  u32 top = 0;
+  u32 left = 0;
+
+  ImagePixelData resized;
+  if (corrected_height <= new_height)
+  {
+    // Handle vertical padding
+
+    const int diff = new_height - std::trunc(corrected_height);
+    top = diff / 2;
+    if (diff % 2 != 0)
+    {
+      // If the difference is odd, we need to have one side be slightly larger
+      top += 1;
+    }
+    resized = Resize(mode, src, new_width, corrected_height);
+  }
+  else
+  {
+    // Handle horizontal padding
+
+    const int diff = new_width - std::trunc(corrected_width);
+    left = diff / 2;
+    if (diff % 2 != 0)
+    {
+      // If the difference is odd, we need to have one side be slightly larger
+      left += 1;
+    }
+    resized = Resize(mode, src, corrected_width, new_height);
+  }
+  CopyTo(resized, result, Rect{Point{0, 0}, resized.width, resized.height},
+         Rect{Point{left, top}, resized.width, resized.height});
+
+  return result;
+}
+
+std::vector<Rect> FindRectRegionsOfNeighboringPixels(const ImagePixelData& src)
+{
+  static const Pixel mask_color{0, 0, 0, 255};
+
+  std::vector<Rect> regions;
+  std::vector<bool> seen_pixels;
+  seen_pixels.resize(src.width * src.height);
```

Comment:
This can be collapsed into the constructor:

```suggestion
  std::vector<bool> seen_pixels(src.width * src.height);
```

or, this could be moved entirely into CalculateNeighboringPixels (see next review comment). It could also be dropped entirely from the interface, given it's not actually used, aside from being stored into in CalculateNeighboringPixels.

Reply:
Unless I'm missing something from your suggestion, I don't think `seen_pixels` can move into `CalculateNeighboringPixels` because it is built up from multiple for loop passes.  The other vector (as you mention below) could be moved to be a return value instead though.

Reply:
Ah OK, I must have missed the read from the vector in the function when doing my initial review. Disregard the latter half of my comment.